cmake/gnu-toolchain.cmake

```diff
@@ -0,0 +1,17 @@
+set(CMAKE_C_COMPILER gcc)
+set(CMAKE_CXX_COMPILER g++)
+
+if(BEMAN_BUILDSYS_SANITIZER STREQUAL "ASan")
```

Comment:
Modeling sanitizers as a build type works better, either that or an entirely distinct toolchain so Thread and Memory can be uniformly applied to all packages. If everything in an address space aren't using msan or tsan the reports they provide are broken, so you have to rebuild and relink the whole world consistently. 
UB sanitizer and address, don't suffer the same problems. 

So something like (not tested!):
```cmake
set(CMAKE_CXX_FLAGS_ASAN
    "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fsanitize=address,undefined,leak"
    CACHE STRING
    "C++ ASAN Flags"
    FORCE
)
```

Also at -O0 there's often no undefined behavior emitted for the sanitizer to see, for the same reasons that debug builds seg fault less often.

Reply:
(optional26 doesn't use the _INIT variables because it's copied from ancient sources before that rule was clarified. Above should be using the *_INIT vars)

Reply:
Thanks for the write-up.

Separate ASAN as build targets seems a bit like overkill for the exemplars use case. 

The design goal here isn't to have a full fledged instrumentation based analysis build system but just a quick hand for "enable all flags for sanitizers". 

Given there's no dependency for exemplar, and the current recommendation for dependency management is to build with dependency's source code instead of including the dependent library at link time. I don't think there's value in complications here.

Reply:
Ah I see where u r coming from, I didn't think about exemplar as a dependency for other libraries (being a dependent) and were only commenting on its use of dependency and a standalone development library / CI test target. 

I think you are right, there should be an ASAN target to produce an ASAN enabled library so someone could link us as a dependency to use. I get what you are talking about. But I think this is more of a package/ export issue, outside of scope for this PR for now and to be honest outside of my skill tree for now.

Again again again, the main motivation here is just to simply CI/ workflow.

Honestly I am tentatively waiting for someone to implement package export, do a quick write up, evaluate it and yonk it over (just like code coverage).

Could we delegate this suggestion to another PR? Let me know if I should add something/ structure this tool chain in anticipation of this feature.

Reply:
> Modeling sanitizers as a build type works better, either that or an entirely distinct toolchain so Thread and Memory can be uniformly applied to all packages.

How so? This isn't obvious for me.

Reply:
Tools like IDEs usually treat selecting the build type as a first class UI operation, where setting custom options is much more buried, so if running sanitizers is a build type, it's much easier to run them reliably that way. 

The alternative of using a distinct toolchain where e.g. memory sanitizer is part of the default build types makes it easier to apply the same options across many packages. That's particularly useful if you're using a package manager and getting pre-built libraries from them if they're available. Mixing thread or memory sanitizer libraries with ones without produces far too many false positives as the sanitizer can't do the tracking for memory initialization or lock acquisition/release in the libraries that aren't instrumented. 

Address and undefined behavior sanitizers don't have that tracing/tracking problem, so they're reasonably accurate if just the code under test is instrumented.